src/app/ClusterData.h

```diff
@@ -0,0 +1,382 @@
+/* Offset from the value of a path that provides the index into the table for that handle
+ * It is 2 since the 0 and 1 are special handles.
+ */
+constexpr AttributePathHandle kHandleTableOffset = 2;
+
+inline bool IsRootAttributePathHandle(AttributePathHandle aHandle)
+{
+    return (aHandle == kRootAttributePathHandle);
+}
+
+inline bool IsNullAttributePathHandle(AttributePathHandle aHandle)
+{
+    return (aHandle == kNullAttributePathHandle);
+}
+
+struct ClusterInfo
+{
+    void Init(void) { this->ClearDirty(); }
+    bool IsDirty(void) { return mDirty; }
+    void SetDirty(void) { mDirty = true; }
+    void ClearDirty(void) { mDirty = false; }
+
+    ClusterDataHandle mClusterDataHandle = 0;
+    bool mDirty                          = false;
+};
+
+/**
+ *  @class ClusterSchemaEngine
+ *
+ *  @brief The schema engine takes schema information associated with a particular Cluster and provides facilities to parse and
+ *         translate that into a form usable by the IM machinery. This includes methods to interpret/query the schema itself and
+ *         methods to help read/write out data to/from TLV given a handle.
+ *
+ *         The schema itself is stored in tabular form, sufficiently described to allow for generic parsing/composition of IM
+ *         paths/data for any given Cluster. These tables are what will be the eventual output of 'code-gen'
+ */
+class ClusterSchemaEngine
+{
+public:
+    /* Provides information about a particular path handle including its parent Attribute schema handle,
+     * its context tag and its name.
+     */
+    struct AttributeInfo
+    {
+        AttributePathHandle mParentHandle;
+        uint8_t mContextTag;
+    };
+
+    /**
+     *  @brief
+     *    The main schema structure that houses the schema information.
+     */
+    struct Schema
+    {
+        chip::ClusterId mClusterId;                   //< The ID of the Cluster.
+        const AttributeInfo * mpAttributeHandleTable; //< A pointer to the schema handle table, which provides parent info and
+                                                      //<context tags for each schema handle.
+        uint32_t mNumAttributePathHandleEntries;      //< The number of schema handles in this Cluster.
+        uint32_t mTreeDepth;                          //< The max depth of this schema.
+    };
+
+    /* While Clusters can have deep nested structures, consumer SDK is only expected to provide
+     * getters/setters for 'leaf' nodes in the schema. If one can visualize a schema as a tree (a directed graph where you can have
+     * at most one parent for any given node) where branches indicate the presence of a nested structure, then this analogy fits in
+     * quite nicely.
+     */
+    class ISetDataDelegate
+    {
+    public:
+        /**
+         * Given a path handle to a leaf node and a TLV reader, set the leaf data in the callee.
+         *
+         * @retval #CHIP_NO_ERROR On success.
```

Comment:
delete these return comments. These effectively say `CHIP_NO_ERROR on succes, other on failure`. I believe that 'set' could get an error (so just saying read failure seems wrong) and return code of 'NO_ERROR' is obvious for anything returning CHIP_ERROR.

same for return comments below.